### ISSUE TITLE
FR: Support large binary files better than Git

### ISSUE THREAD

**Issue by PhilipMetzger** [martinvonz/jj#2865]
**Is your feature request related to a problem? Please describe.**

In certain industries like Games and Audio it'd be useful if `jj` could store Audio  and Art efficiently. Currently we'd need to hack it into the Git backend, but if we ever have a native backend it should be simplified. Git LFS is not an alternative.

**Describe the solution you'd like**

We should support binary files with some kind of [CDC](https://en.wikipedia.org/wiki/Change_data_capture) mechanism and a special object tag.

**Additional context**
Discord conversations throughout the last 1.5 years. And the [FastCDC paper][fastcdc] for those who are interested. 

I found only one conversation [here][dc1].

[dc1]: https://discord.com/channels/968932220549103686/1168642196199706776/1168645704068120656
[fastcdc]: https://www.usenix.org/system/files/conference/atc16/atc16-paper-xia.pdf

**Comment by ilyagr** [comment]
Naively, the data model of Restic (https://restic.net/) seems like it would work for our usecase. In fact, we could just have a restic repository as a backend, which would make binary file support work immediately with many storage backends (e.g. Google Cloud, Backblaze B2). Each commit could correspond to a Restic snapshot.

This assumes that Restic's deduplication strategy is good enough and works well for pulling up files based on their hashes. I am not an expert on it, but I also haven't heard many complaints. I'm not sure whether it uses the same CDC mechanism as Philip mentioned.

Another equally popular backup tool is https://borgbackup.readthedocs.io/en/stable/. I don't have a strong opinion between the two; I'm highlighting Restic because I've tried it and its community seems pleasant.

Restic is written in Go, but there also exists https://github.com/rustic-rs/rustic.

**Comment by necauqua** [comment]
Reading on both of those, borg seems more interesting

Particularly, this part stood out for me (I mean, it's out there almost immediately on the main page heh)
> Deduplication based on content-defined chunking is used to reduce the number of bytes stored: each file is split into a number of variable length chunks and only chunks that have never been seen before are added to the repository.
>
> A chunk is considered duplicate if its id_hash value is identical. A cryptographically strong hash or MAC function is used as id_hash, e.g. (hmac-)sha256.
>
> To deduplicate, all the chunks in the same repository are considered, no matter whether they come from different machines, from previous backups, from the same backup or even from the same single file.
>
> Compared to other deduplication approaches, this method does NOT depend on:
> - file/directory names staying the same: So you can move your stuff around without killing the deduplication, even between machines sharing a repo.
> - complete files or time stamps staying the same: If a big file changes a little, only a few new chunks need to be stored - this is great for VMs or raw disks.
> - The absolute position of a data chunk inside a file: Stuff may get shifted and will still be found by the deduplication algorithm.

**Comment by ilyagr** [comment]
For Rsestic, there's a lot of info here: https://restic.readthedocs.io/en/latest/100_references.html#design. I'm hoping to look at it more carefully one day.

**Comment by PhilipMetzger** [comment]
First of sorry for the late response here. 

> Naively, the data model of Restic (https://restic.net/) seems like it would work for our usecase. In fact, we could just have a restic repository as a backend, which would make binary file support work immediately with many storage backends (e.g. AWS, Backblaze B2, anything supported by `rclone`). Each commit could correspond to a Restic snapshot.
> 
> One possible limitation of Restic is that I'm not sure it's very good at pushing/pulling snapshots between the repositories. It has this functionality, but I'm not sure whether it's polished and optimized for our use-case.

A restic backend could solve the use-case but doesn't actually achieve what I want. I want a better support for large files natively and as such, probably involves some hacks on the Git model (the object tag/storage strategy, I hinted at). 

For the native repo we just should use a CDC mechanism anyway.  

> This assumes that Restic's deduplication strategy is good enough and works well for pulling up files based on their hashes. I am not an expert on it, but I also haven't heard many complaints. I'm not sure whether it uses the same CDC mechanism as Philip mentioned.

It probably won't solve the use-case for this bug, but it should be a good trial run for an alternative backend.

**Comment by Timmmm** [comment]
I think good large file support isn't just about efficient diffing. You also want:

1. Not to have to worry about large files from ancient commits cluttering up your repo for eternity.
2. To decouple the storage mechanism from the file content (LFS fails miserably at this). Ideally you would be able to do this retroactively.

I always thought the way LFS worked was dumb. The actual commit shouldn't care about the storage mechanism. It only cares about the content so it should always just store the hash of the blob and then let something else worry about how it is stored.

It would be much better if it was something like `git notes` which can be added on to a commit after a fact that then said "Oh by the way, this blob - 03828fe3d3 - don't download it from the repo; get it from this file server instead. And don't download it by default; it's ancient and the user probably doesn't want it."

**Comment by asayers** [comment]
You guys may already be aware of it, but some highly relevant prior art is [bup](https://bup.github.io/), which adds CDC support to the git repo format.

There's a [fun explanation of how it works](https://github.com/bup/bup/blob/main/DESIGN.md#handling-large-files-cmdsplit-hashsplitsplit_to_blob_or_tree).  But the important observation is to use the rolling hash not only to start new chunks, but also to start new subtrees.  This lets you efficiently share runs of many identical chunks. (I believe this effectively gives you a [Merkle search tree](https://inria.hal.science/hal-02303490/document).  Independent discovery by bup and Inria?)

**Comment by tv42** [comment]
My use case:

I have a personal project [plop](https://github.com/bazil/plop) that writes large files split with rolling checksums into object stores and replaces them with a symlink to FUSE mount that serves the content, read-only, roughly deduplicated, compressed, encrypted. The content is never fully copied back out, just served through FUSE chunk by chunk and cached in RAM. `git clone` is near-instantaneous with no shallow/partial trickery. It can talk to multiple object stores concurrently, so recent files can be kept local etc. I like it more than Microsoft's LFS.

If I am about to "git stage" a file matching a configured glob, I want to first run my tool on the file, and then act on whatever is left of the file. The git tree or index should only see the symlink/whatever my tool replaces the file with, the original file content should never enter into the git store.

(Git's strict stdin-to-stdout filters can't express the "replace with a symlink" part.)

As of today, I need to manually remember to `plop add PATH...`, and jj's oversize protection prevents me from forgetting (for repos with big files, `snapshot.auto-track` off avoids lots of oversize complaints). Finger memory is still stuck at `plop add foo && git add foo`.

**Comment by Timmmm** [comment]
I don't think the solution should store the binary files outside of the repo like LFS. That's just a hack.

I think an ideal solution would semantically include *all* files in the same history, but provide tools to avoid having to download them all. Git *sort of* does that with blob filters, sparse checkouts, `--depth 1` and all of that, but it's very unergonomic (quelle surprise).

My ideal solution would be something where you store the large files in Git (in an efficient way), but then *later* you can go back and mark large files with things like "by default don't download this when cloning" and "get this from S3 instead".

**Comment by tv42** [comment]
@Timmmm Agreed in the sense that storing large content outside of the repository is hack, for sure. But it's a great trick around limitations like max repository size (my tool can e.g. fill up multiple NASses in your basement one at a time), local caches (my tool does happy-eyeballs style concurrent requests, and you can e.g. set a 100ms delay before cloud object storage is attempted and keep most requests local), etc. If you take the absolutist stance, you end up having to solve two mostly disjoint problems in one architecture. Of course, you can reformulate the repository storage to work over a generic object store to unify the two, but then you're throwing away the Git code, and need to be really careful not to lose the pack optimizations.

**Comment by Timmmm** [comment]
> then you're throwing away the Git code

Kind of the point! I would be happy with a solution that works well with JJ but less well with Git. For example, storing all of the binary files in the repo exactly like `git add` does, but adding extra metadata so that JJ can be smarter about when it downloads that data and from where.

**Comment by nickchomey** [comment]
Fwiw, I've been using rustic for a while and it works great. 

This is a similar tool that came out recently, which might be worth looking at. https://github.com/PlakarKorp/plakar

**Comment by rahuldave** [comment]
Am thinking that another use for this is agent conversation logs. These can grow quite large in what i call session-oriented programming where i am learning/exploring, rather than a very well scoped prd.  My idea is to use `jj new` liberally, probably at the start/end of each user-prompt hook (i think codex only provides start) or at the end of a patch tool call (in codex) to have undescribed commits point to shorter increases in the agent log file (say jsonl). This gives me pretty fine-grained repo-history to agent log matching. But having something that captures the log state as a prt of the commit but without it being in the repo itself would be great for provenance...

**Comment by LoganDark** [comment]
> Am thinking that another use for this is agent conversation logs. These can grow quite large in what i call session-oriented programming where i am learning/exploring, rather than a very well scoped prd. My idea is to use `jj new` liberally, probably at the start/end of each user-prompt hook (i think codex only provides start) or at the end of a patch tool call (in codex) to have undescribed commits point to shorter increases in the agent log file (say jsonl). This gives me pretty fine-grained repo-history to agent log matching. But having something that captures the log state as a prt of the commit but without it being in the repo itself would be great for provenance...

I feel like this would not be a good fit for text files. This is for binary files where splitting/diffing based on ASCII newline is inappropriate.

If you want to keep track of the repo state over time, then perhaps the [operation log](https://docs.jj-vcs.dev/latest/operation-log) may help.
